lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -252,6 +254,123 @@ void FIRRTLTypesLowering::visitDecl(InstanceOp op) {
   opsToRemove.push_back(op);
 }
 
+void FIRRTLTypesLowering::visitDecl(MemOp op) {
+
+  auto type = op.getDataTypeOrNull();
+  // Bail out if the memory is empty
+  if (!type)
+    return;
+
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(op.getDataTypeOrNull(), "", false, fieldTypes);
+
+  auto maybeFlip = [](Type a) -> std::tuple<BundleType, bool> {
+    return TypeSwitch<Type, std::tuple<BundleType, bool>>(a)
+        .Case<FlipType>([](FlipType b) {
+          return std::make_tuple(b.getElementType().cast<BundleType>(), true);
+        })
+        .Default([](Type b) {
+          return std::make_tuple(b.cast<BundleType>(), false);
+        });
+  };
+
+  DenseMap<StringRef, Value> newWires;
+  for (auto field : fieldTypes) {
+    // Create the new type for the memory. Copy every field, but
+    // update the "data" field to be the groundn type and the "mask"
+    // field to be a uint<1> type.
+    SmallVector<Type, 8> resultTypes;
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      SmallVector<BundleType::BundleElement, 5> resultFields;
+
+      BundleType portType;
+      bool isFlipped;
+      std::tie(portType, isFlipped) = maybeFlip(op.getResult(i).getType());
+
+      for (auto elt : portType.getElements())
+        resultFields.push_back(
+            StringSwitch<BundleType::BundleElement>(elt.name)
+                .Cases("data", "rdata", "wdata", {elt.name, field.type})
+                .Cases("mask", "wmask",
+                       {elt.name, UIntType::get(builder->getContext(), 1)})
+                .Default(elt));
+
+      if (isFlipped) {
+        resultTypes.push_back(FlipType::get(
+            BundleType::get(resultFields, builder->getContext())));
+        continue;
+      }
+
+      resultTypes.push_back(
+          BundleType::get(resultFields, builder->getContext()));
+    }
+
+    // Construct the new memory for this flattened field
+    auto newName = op.name().getValue().str() + field.suffix;
+    auto newMem = builder->create<MemOp>(
+        resultTypes, op.readLatency(), op.writeLatency(), op.depth(), op.ruw(),
+        op.portNames(), builder->getStringAttr(newName));
+
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+
+      BundleType portType;
+      bool isFlipped = false;
+      std::tie(portType, isFlipped) = maybeFlip(op.getResult(i).getType());
+
+      for (auto elt : portType.getElements()) {
+        switch (StringSwitch<int>(elt.name)
+                    .Cases("clk", "en", "addr", "wmode", 0)
+                    .Cases("mask", "wmask", 1)
+                    .Default(/* "data", "rdata", "wdata", */ 2)) {
```

Comment:
Maybe make this an explicit `Cases`, and have a `Default` that leads to an error/assertion?

Reply:
I thought about this... However, if the memory contains a field that isn't one of these types, the type of the port/memory is somehow wrong. This would probably be better handled with a `firrtl::MemOp` verifier. I'll open an issue about this.